### PR TITLE
Add deploy task for post move survey email

### PR DESCRIPTION
**Pull request by mkrump**
## Description

Supports https://github.com/transcom/ppp-infra/pull/604

Add deploy task for post move survey email

## References

* [Pivotal story](https://www.pivotaltracker.com/n/projects/2181745/stories/168035083) for this change

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -97,6 +97,18 @@ commands:
             DB_SSL_ROOT_CERT: /bin/rds-combined-ca-bundle.pem
           no_output_timeout: 20m
       - announce_failure
+#      TODO remove --offset-days from args before merge to default back to 15 days.
+      - deploy:
+          name: Deploy post move email survey task service
+          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command send-post-move-survey-email --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2 --offset-days 1"
```
Comment (mr337) [inline .circleci/config.yml]: More curious than anything, where is the `--offset-days` coming from? Can't find it in the `do-exclusively` or `ecs-deploy-task-container` program.

Comment (mkrump) [reply .circleci/config.yml]: possible that might not work as structured, but was attempting to pass a command line flag to binary.

Comment (chrisgilmerproj) [reply .circleci/config.yml]: It's correct, basically you're passing through an argument to the Task Definition which is running the Docker Container which itself is running the binary. It's that binary that has `--offset-days`.

**Comment by chrisgilmerproj** [comment]
There's no validation done here when setting up the notification sender to see if it can send email: https://github.com/transcom/mymove/blob/master/pkg/notifications/notification_sender.go#L65-L76

`CheckEmailFlags` is supposed to do at least a minimum of configuration checking. But would be nice if we had something similar to the DB Ping() that we use to make sure we can talk to Postgres before connecting. A good example might be `aws ses get-account-sending-enabled ...`.

**Comment by mkrump** [comment]
> But its also unclear why the client didn't throw an error when we configured it without a domain or region. We should figure that out too.

Added `CheckEmailFlags(v)`. I think that there was no error b/c it's getting from the local environment (at least for me I see).

```
(base) Projects/mymove - [mk-168035083-deploy-ecs-email-task●] » env | grep -i aws_ses
AWS_SES_REGION=us-west-2
AWS_SES_DOMAIN=devlocal.dp3.us
```

**Comment by chrisgilmerproj** [comment]
> In [ff2a310](https://github.com/transcom/mymove/commit/ff2a3100299ddb2ac7d5871c500dcb0dba452a04) added a call to `get-account-sending-enabled`. Right now an error results in a `log.Fatal` just wanted to confirm that is the behavior that you are expecting.

Yes! This is exactly what we need. If we expect to use email we need to make sure its enabled for use.  Thanks for adding this.

**Comment by mkrump** [comment]
@chrisgilmerproj linda got an email over the weekend. if everything else looks ok, should be good to go then (after updating the offset). Any idea why the deploy to experimental failed though? Should I just re-run to make sure that it deploys successfully?

**Comment by chrisgilmerproj** [comment]
> @chrisgilmerproj linda got an email over the weekend. if everything else looks ok, should be good to go then (after updating the offset). Any idea why the deploy to experimental failed though? Should I just re-run to make sure that it deploys successfully?

Link me to the failure and I'll look. I'm thinking we still have a little more work to do on this utility before we want to send it out to Production. In particular we ought to make sure that it has exponential backoff in case we get rate limited. I'm also worried about the tooling we have around sending/resending if there are problems.

**Comment by mkrump** [comment]
Wanted to clarify any additional changes needed before merge.

> I'm thinking we still have a little more work to do on this utility before we want to send it out to Production. In particular we ought to make sure that it has exponential backoff in case we get rate limited. I'm also worried about the tooling we have around sending/resending if there are problems.

For the rate limiting, I thought that we'd agreed that a [future story](https://www.pivotaltracker.com/n/projects/2181745/stories/168305366) was sufficient. We probably don't even need to get that complicated. A small delay would probably suffice since the limit is `80 emails/second` and it looks like it needs to be sustained, not just quickly hit https://docs.aws.amazon.com/ses/latest/DeveloperGuide/manage-sending-limits.html?icmpid=docs_ses_console. 

> I'm also worried about the tooling we have around sending/resending if there are problems.

I wasn't sure if looking for a change here. But I think this probably applies to all emails that are sent at the moment though, not just these.

**Comment by chrisgilmerproj** [comment]
> For the rate limiting, I thought that we'd agreed that a [future story](https://www.pivotaltracker.com/n/projects/2181745/stories/168305366) was sufficient. We probably don't even need to get that complicated. A small delay would probably suffice since the limit is `80 emails/second` and it looks like it needs to be sustained, not just quickly hit https://docs.aws.amazon.com/ses/latest/DeveloperGuide/manage-sending-limits.html?icmpid=docs_ses_console.

I see the confusion. I expected a second story would be fine if we finished it before going to production. I think you heard it as saying we could push to production and then revisit. I'm nervous that it will get dropped if we do that and I'd rather have it in hand ahead of time. I'm open to opinions on that though (whether it will get done in a reasonable time frame).
 
> I wasn't sure if looking for a change here. But I think this probably applies to all emails that are sent at the moment though, not just these.

There's no real admin for managing emails. So if a bad email goes out we don't really have re-call. We don't have the ability to re-send. There's pretty much no dials or switches that infra can use or app devs can use to manage the emails. It feels a bit risky to launch without something that would let us manage the emails that get picked up by the scheduled task.  If you disagree that's fine, and you're right that its a larger problem, but we are putting ourselves in a tougher situation without having something in hand.

**Comment by chrisgilmerproj** [comment]
Oh, the other thing is I was hoping to combine the fuel task and the email task into a single binary with sub-commands. Would that be in this PR?

**Comment by jim** [comment]
> Oh, the other thing is I was hoping to combine the fuel task and the email task into a single binary with sub-commands. Would that be in this PR?

I think that needs to be in another story.

> There's no real admin for managing emails. So if a bad email goes out we don't really have re-call. We don't have the ability to re-send. There's pretty much no dials or switches that infra can use or app devs can use to manage the emails. It feels a bit risky to launch without something that would let us manage the emails that get picked up by the scheduled task. If you disagree that's fine, and you're right that its a larger problem, but we are putting ourselves in a tougher situation without having something in hand.

This is true, but we're already sending emails in other parts of the app that this is also true for. It sounds like we need to plan out some additional work to have better email infrastructure, but I don't think we should block this story on that.

The emails in question here are survey emails that are send after the move is completed, so a delivery failure is an acceptable risk in this case.

**Comment by jim** [comment]
> In particular we ought to make sure that it has exponential backoff in case we get rate limited.

This is something we can iterate on going forward. We have a very small number of users in the system right now and it's unlikely we're going to hit such a rate limit in the near future.

**Comment by chrisgilmerproj** [comment]
@jim @mkrump - we can move forward on this but I don't consider this to be "done". There's clearly work that needs to be done on the email system that we're discovering now that we're modifying it and I fear it won't actually be addressed. Also the scheduled tasks (fuel, email) really need to be combined into a binary so we can start looking at doing acceptance testing before deploys as currently there is no test that validates the binaries won't fail in production. 

I'll give my approval I just don't want us to never come back to these issues as they will absolutely bite us in the future.

**Comment by mkrump** [comment]
I can do this.

> Also the scheduled tasks (fuel, email) really need to be combined into a binary so we can start looking at doing acceptance testing before deploys as currently there is no test that validates the binaries won't fail in production.

Is there a middle ground on the other stuff? Like the concern about rate limiting? SES says our burst rate limit is 80 per 1 sec (or an email every 12.5ms). Would adding a sleep to the loop for 20 ms help alleviate that concern? 

The admin for managing emails really does feel like something separate that we need to probably talk to product about (unless this is more straightforward than i'm thinking).

**Comment by chrisgilmerproj** [comment]
> Is there a middle ground on the other stuff? Like the concern about rate limiting? SES says our burst rate limit is 80 per 1 sec (or an email every 12.5ms). Would adding a sleep to the loop for 20 ms help alleviate that concern?

Yeah, that would be fine. Anything that keeps us out of the danger zone for rate limiting will help us. My real concern there is that the error we hit when we do get rate limited will be inscrutable so we might waste time trying to figure out what's happening.

> The admin for managing emails really does feel like something separate that we need to probably talk to product about (unless this is more straightforward than i'm thinking).

I'm ok having the admin in another ticket. I just don't know what we'll do about resending because honestly I'm not even sure how we capture information on send failures to know something went wrong. It just feels like we're building a system where we have no real visibility into our success/failure just anecdotal evidence that it worked while we were looking at it. Kind of like a Schrodinger's email box, works when we watch, no idea what it does when we're not looking.

This makes me think that we need to be capturing email send success/failure stats somewhere with some kind of actionable alert. And if we have something like that we need to write it up so the on-call person knows what to do if we start failing a lot.

**Comment by mkrump** [comment]
@chrisgilmerproj and @jim. 

I added a sleep 9db08a3382ba4488cd9fb50712370315011264d6 to address the concerns about possible rate limiting.

Also, I took a stab at combining under single binary c52b61c52c15c678dc74e0d1d606e79bcb8dbace. This feels like a pretty big change to make at the end of the pr, so could I get some careful eyes on that. I especially don't want to mess up anything with existing save-fuel-price data. 

Experimental was occupied at the moment, but was hoping when that frees up, that we could do a final round of manual testing and then get this merged in tomorrow (if possible).

File: .circleci/config.yml [hunk 2]
```diff
@@ -87,8 +87,20 @@ commands:
...
+            DB_USER: ecs_user
+            DB_NAME: app
+            DB_SSL_MODE: verify-full
+            DB_SSL_ROOT_CERT: /bin/rds-combined-ca-bundle.pem
+          no_output_timeout: 20m
+      - announce_failure
+#      TODO remove --offset-days from args before merge to default back to 15 days.
+      - deploy:
+          name: Deploy post move email survey task service
+          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command ecs-tasks send-post-move-survey-email --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2 --email-backend=ses --offset-days 0"
```
Comment (chrisgilmerproj) [inline .circleci/config.yml]:
```suggestion
          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command "ecs-tasks send-post-move-survey-email" --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2 --email-backend=ses --offset-days 0"
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -87,8 +87,20 @@ commands:
       - attach_workspace:
           at: bin
       - deploy:
-          name: Deploy task service
-          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command save-fuel-price-data --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2"
+          name: Deploy fuel price data task service
+          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command ecs-tasks save-fuel-price-data --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2"
```
Comment (chrisgilmerproj) [inline .circleci/config.yml]: Needs quotes for the `--command`:

```suggestion
          command: scripts/do-exclusively --job-name ${CIRCLE_JOB} bin/ecs-deploy-task-container --aws-account-id ${AWS_ACCOUNT_ID} --aws-region ${AWS_DEFAULT_REGION} --service app --environment ${APP_ENVIRONMENT} --repository-name app-tasks --image-tag git-${CIRCLE_SHA1} --command "ecs-tasks save-fuel-price-data" --command-args "--db-iam --db-iam-role arn:aws:iam::923914045601:role/ecs-rds-role-app-${APP_ENVIRONMENT} --db-region us-west-2"
```

File: cmd/ecs-deploy-task-container/main.go [hunk 4]
```diff
@@ -26,7 +26,7 @@ var services = []string{"app"}
 var environments = []string{"prod", "staging", "experimental"}
 
 // Commands should be the name of the binary found in the /bin directory in the container
-var commands = []string{"save-fuel-price-data"}
+var commands = []string{"ecs-tasks"}
```
Comment (chrisgilmerproj) [inline cmd/ecs-deploy-task-container/main.go]: I bet this has a problem and we'll need to modify to:

```suggestion
var commands = []string{"ecs-tasks save-fuel-price-data", "ecs-tasks send-post-move-survey-email"}
```

File: cmd/ecs-tasks/version.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+func versionFunction(cmd *cobra.Command, args []string) error {
+	str, err := json.Marshal(map[string]interface{}{
+		"gitBranch": gitBranch,
+		"gitCommit": gitCommit,
```
Comment (chrisgilmerproj) [inline cmd/ecs-tasks/version.go]: To get these I think we need to build the binary with the branch and commit as data we pass in to `go build`

File: cmd/ecs-deploy-task-container/main.go (hunk 4 above)
Comment (mkrump) [reply cmd/ecs-deploy-task-container/main.go]: ok. yeah wasn't sure what to do here.

File: .circleci/config.yml (hunk 3 above)
Comment (mkrump) [reply .circleci/config.yml]: 👍

File: cmd/ecs-tasks/version.go (hunk 5 above)
Comment (mkrump) [reply cmd/ecs-tasks/version.go]: i copied this from the existing milmove binary (so not sure if we even need this here). but if so, i'm not immediately seeing where we do this for milmove. could you point me to that?

Comment (mkrump) [reply cmd/ecs-tasks/version.go]: Add to build in https://github.com/transcom/mymove/pull/2655/commits/9173728713e43d550f2bf0d6f21392f9255eed27. Tested with `version` and looked like capturing version info now.

**Comment by mkrump** [comment]
reverted the experimental deploy and changed offset back to 15 days, so i think on monday this is ready to merge.

File: cmd/milmove-tasks/main.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		RunE:         sendPostMoveSurvey,
+		SilenceUsage: true,
+	}
+	initPostMoveSurveyFlags(sendPostMoveSurveyCommand.Flags())
+	root.AddCommand(sendPostMoveSurveyCommand)
+
+	completionCommand := &cobra.Command{
+		Use:   "completion",
+		Short: "Generates bash completion scripts",
+		Long:  "To install completion scripts run:\n\nmilmove completion > /usr/local/etc/bash_completion.d/milmove",
```
Comment (chrisgilmerproj) [inline cmd/milmove-tasks/main.go]:
```suggestion
		Long:  "To install completion scripts run:\n\nmilmove-tasks completion > /usr/local/etc/bash_completion.d/milmove-tasks",
```

Comment (mkrump) [reply cmd/milmove-tasks/main.go]: 👍 updated.

**Pull request merged**